### ISSUE TITLE
Memories app causes Nextcloud cron to use excessive amount of resources (oc_filecache queries)

### ISSUE THREAD

**Issue by Hermanio** [pulsejet/memories#1315]
### Describe the bug

I noticed that my Nextcloud server uses a lot of CPU periodically, and the timing correlates with running the Nextcloud cron task.

I enabled logging all queries on my PostgreSQL instance and noticed that Nextcloud cron task runs a lot of queries against `oc_filecache` table:
```
2024-10-15 18:50:31.782 UTC [40] LOG:  execute <unnamed>: SELECT "filecache"."fileid", "storage", "path", "path_hash", "filecache"."parent", "filecache"."name", "mimetype", "mimepart", "size", "mtime", "storage_mtime", "encrypted", "etag", "filecache"."permissions", "checksum", "unencrypted_size", "metadata_etag", "creation_time", "upload_time", "meta"."json" AS "meta_json", "meta"."sync_token" AS "meta_sync_token" FROM "oc_filecache" "filecache" LEFT JOIN "oc_filecache_extended" "fe" ON "filecache"."fileid" = "fe"."fileid" LEFT JOIN "oc_files_metadata" "meta" ON "filecache"."fileid" = "meta"."file_id" WHERE "filecache"."parent" = $1 ORDER BY "name" ASC
```

This results in a high CPU activity on both Nextcloud and the database container (PostgreSQL 16). 

My instance is not very large, about 400K files in total, images make up a fraction of that (10-25K probably). 
I've already truncated the `oc_filecache` table, reindexed the table in Postgres, and done a full rescan via `occ`, but that did not improve the situation.

I've done all the optimizations mentioned in the issue template, even before using Memories.
Transcoding is enabled. Recognize is disabled.

After disabling Memories app within Nextcloud, the cron task uses a reasonable amount of CPU.

A quick code search over this repo suggests use of the filecache table so I suspect that the load is caused by the Memories app. I've seen similar reports of other people struggling with similar issues on the filecache table, but I did not notice any relations to a specific app like Memories.

I tried searching existing issues for similar topics but could not find one. 

### Steps To Reproduce

I don't have the time to test this out on a fully clean Nextcloud system, but I reckon it can be tested like this:
1. set up a clean Nextcloud instance (Docker, Postgres)
2. install and enable Memories
3. create a large number of files (1+ million to be safe), a good chunk of them images/videos
4. set Nextcloud cron task to be run every 5 minutes
5. observe CPU and container resource usage over time

### Platform

```markdown
Platform:
Nextcloud Hub 8 (29.0.7) (official Docker image)
Memories 7.4.1
PostgreSQL 16 (official Docker image)
```

### Screenshots

Attached is the CPU usage graph of my server, with Memories app enabled and disabled.

![Screenshot from 2024-10-16 19-37-59](https://github.com/user-attachments/assets/40d70e5e-d731-41e6-8d12-679b63e3b051)

I have not had this run for a very long time, but the effect is noticeable already.

### Additional context

No relevant errors in Nextcloud logs.

**Comment by pihomeserver** [comment]
I have the same issue and after searching i got the same conclusion about the cron job CPU usage getting high

![Image](https://github.com/user-attachments/assets/3ffdfd27-6f9b-4e25-864c-50ffd6252966)

Looking at cron job table it seems to be related to the following job with 248 seconds of execution time (all others are 0 or near 0)
```
 OCA\Memories\Cron\IndexJob                                        |                248
```

**Comment by pihomeserver** [comment]
@pulsejet hi. my 4 S3 buckets have a total of 33000 files. If i could limit preview to only the one i want then it would reduce to 7300 photos.

**Comment by pulsejet** [comment]
Hmm that's basically nothing then, there are file folks using memories for 1m+ files. This could likely have something to do with either nextcloud's s3 implementation or something went wrong with the db

**Comment by Hermanio** [comment]
Could there be a link with the default background indexing setting? 

> The EXIF indexes are built and checked in a periodic background task. Be careful when selecting anything other than automatic indexing. For example, setting the indexing to only timeline folders may cause delays before media becomes available to users, since the user configures the timeline only after logging in. 

![Image](https://github.com/user-attachments/assets/a60eea03-097d-47df-9ee7-f0d896674d22)

I think that disabling the automatic indexing and setting up a separate scheduled cronjob that calls `occ memories:index` can be considered a valid workaround to this issue if we can't find a way to make the background indexing less resource intensive. 

I already have a daily Nextcloud maintenance script running, and I don't need the latest images to be shown within the Memories app, so for me this workaround will likely do the trick. 

Would be good to understand if the default behaviour can be improved, but that's up to the maintainers of this app to decide. I reckon that it's a fun optimization task and I'd love to help if I had the time to do some thorough digging.

I'll test this change out and report back with the results once I have enough data (and once the reindexing finishes).

**Comment by pulsejet** [comment]
Yes, it's a valid workaround. Files are indexed on upload but the background job covers some corner cases when it doesn't. Ideally the background job is very fast (obviously there's a bug somewhere) 

One way to have both might be to have a nightly Cron that does `occ memories:index`. Not strictly necessary.

**Comment by comk22** [comment]
I was messing around for days because of this. "Glad" that others also have this problem. This can't be a normal behavior because NextCloud usually doesn't have these kinds of CPU loads when nothing is really happening. Never had this issue with memories before too. What is confusing is the blue warning "Automatic Indexing status: Indexing process stopped before completion. Will continue on next run". This also shows when I manually --force an index run. I can also see in the console that the indexing is completed.

I also can't really pinpoint when this started. Only disabling the automatic indexing will solve the problem.

**Comment by rotdrop** [comment]
I actually have the same issue. Something like 15k files, the admin settings always shows "Indexing process stopped before completion. Will continue on next run". I did run `./occ memories:index`, just after  the installation. The cron-job takes up to 5 minutes. I guess I'll follow @Hermanio and disable the automatic background job altogether, it just eats too much resources.

**Comment by vermeeren** [comment]
I am also having the problem and believe a very large number of directories may be at fault. I have a collection of audio and video that does not necessarily need indexing by Memories which is around 11TB in total. It consists of around 210k files but more importantly 25k directories with a structure that is in some cases several directories deep.

Running `occ memories:index` once everything is already indexed is quite slow with a lot of `Indexing folder ...` output. For now I added the `.nomemories` file to the root folder of this collection and with that the cron only takes like 15 seconds to check the photos etc I like to see in the timeline.

@pulsejet Could it be that although indexed files are efficiently cached directories are not currently?

Thanks!

**Comment by pulsejet** [comment]
Yeah having lots of directories or lots of files in one directory could be slow. This could be optimized further but it's hard to do safely. 

Just turn off the cron job from the admin panel, it's only needed in edge cases anyway. If you want you can set up a nightly cron job to do this externally.

**Comment by vermeeren** [comment]
@pulsejet Thanks for the info, indeed turning off the cron makes sense to me.

Random observation that might be useful: It seems like Nextcloud directory modification timestamp changes propagate all the way up to the root folder, at least in the web UI. If I change something in `Pictures/Photos/Upload/...` The very `Pictures` in the root folder also updates the modification time. It seems like with this an efficient caching approach is possible rather easily:

1. When starting indexation (cron or manual) get the current timestamp.
2. Retrieve previous indexation timestamp (when it started).
3. Only recurse inside directories with modification timestamp newer than previous index timestamp.
4. Upon completion write the index start timestamp of current index execution.

With this only a single global last-index-start timestamp needs to be stored and the index job will only check directories that contain changes (directly or indirectly through subdirs). In a large collection 99% of directories will typically not change so skipping 99% of directory recursion and file checking inside those directories should be a massive speed increase.

Thoughts on this? Don't have in-depth experience with the code base so just an idea at this point.

Cheers

---

Edit: Another idea, in addition to the above, not really needed because a single `memories:index` manually will solve this too, but possibly interesting anyway.

There is also the case where execution does not finish within the cron time limit. For this it might be possible to store the current position of the scan and continue from there on the next cron execution instead of from the beginning. This way it will eventually complete.

For a split-scan like this the index-start-timestamp should be the one from the first job, as that is the one that started from nothing. So if it would take 5 days to complete initial scan the timestamp stored at the end is the one from 5 days ago. The next job then only needs to check dirs with changes since 5 days ago.

**Comment by pulsejet** [comment]
df4ffaafb018f9a5ccd200e74b78ae429aecbee3
